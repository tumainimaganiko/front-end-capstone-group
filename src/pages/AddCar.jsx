import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCar } from '@fortawesome/free-solid-svg-icons';
import { addCar } from '../redux/cars/carsSlice';
import { fetchModels } from '../redux/cars/modelsSlice';
import Loader from '../components/loader/Loader';

const AddCarPage = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const models = useSelector((state) => state.models.models);
  const isLoading = useSelector((state) => state.cars.isLoading);
  const error = useSelector((state) => state.cars.error);
  const cars = useSelector((state) => state.cars.cars);

  const [isImageValid, setIsImageValid] = useState(true);
  const [isValidPlate, setIsValidPlate] = useState(true);

  const [formData, setFormData] = useState({
    name: '',
    plate_number: '',
    image: null,
    price: '',
    status: true,
    city: '',
    model_id: '',
  });

  useEffect(() => {
    dispatch(fetchModels());
  }, [dispatch]);

  const handleSubmit = (e) => {
    e.preventDefault();
    const data = new FormData();
    const {
      name, plate_number, price, city, model_id,
    } = formData;
    data.append('name', name);
    data.append('plate_number', plate_number);
    data.append('image', e.target.image.files[0]);
    data.append('price', price);
    data.append('city', city);
    data.append('model_id', model_id);
    data.append('status', true);
    dispatch(addCar({ car: data }));
    if (!isLoading && !error) {
      setFormData({
        name: '',
        plate_number: '',
        image: null,
        price: '',
        status: true,
        city: '',
        model_id: '',
      });
      navigate('/cars');
    }
  };

  if (isLoading) {
    return <Loader />;
  }

  const handleInputChange = (field, value) => {
    setFormData((prevData) => ({
      ...prevData,
      [field]: value,
    }));
  };

  const handleModelChange = (e) => {
    const model_id = parseInt(e.target.value, 10);
    setFormData((prevData) => ({
      ...prevData,
      model_id,
    }));
  };

  const handleImageChange = (field, file) => {
    setFormData((prevData) => ({
      ...prevData,
      [field]: file,
    }));
  };

  const validatePlate = () => {
    const plate = formData.plate_number;
    const car = cars.find((car) => car.plate_number === plate);
    if (car) {
      setIsValidPlate(false);
    } else {
      setIsValidPlate(true);
    }
  };

  const handlePlateBlur = () => {
    validatePlate();
  };

  const validateImage = () => {
    if (formData.image !== null && formData.image.type.startsWith('image/')) {
      setIsImageValid(true);
    } else {
      setIsImageValid(false);
    }
  };
  const handleImageBlur = () => {
    validateImage();
  };

  return (
    <section className="p-8">
      <div className="max-w-md mx-auto">
        <h1 className="text-2xl font-bold mb-4">Add New Car</h1>
        <form id="Add-car-form" onSubmit={handleSubmit}>
          <div className="mb-4">
            <label htmlFor="name" className="block font-medium">
              Car Name:
            </label>
            <input
              type="text"
              id="name"
              className="border border-gray-300 rounded w-full "
              onChange={(e) => handleInputChange('name', e.target.value)}
              value={formData.name}
              required
            />
          </div>

          <div className="mb-4">
            <label htmlFor="model_id" className="block font-medium">
              Select Model:
            </label>
            <select
              id="model_id"
              className="border border-gray-300 rounded w-full"
              onChange={handleModelChange}
              value={formData.model_id}
              required
            >
              <option value="">-- Select a Model --</option>
              {models.map((model) => (
                <option key={model.id} value={model.id}>
                  {model.name}
                </option>
              ))}
            </select>
          </div>

          <div className="mb-4">
            <label htmlFor="carDetails" className="block font-medium">
              Car Plate:
            </label>
            <input
              type="text"
              id="plate_number"
              className={`border border-gray-300 rounded w-full ${!isValidPlate ? 'border-red-500' : ''}`}
              onChange={(e) => handleInputChange('plate_number', e.target.value)}
              onBlur={handlePlateBlur}
              value={formData.plate_number}
              required
            />
            {!isValidPlate && (
              <p className="text-red-500 text-sm">The plate number should be unique</p>
            )}
          </div>

          <div className="mb-4">
            <label htmlFor="carPhoto" className="block font-medium">
              Car Photo:
            </label>
            {/* <input
              type="url"
              id="carPhoto"
              className={`border border-gray-300 rounded w-full ${
                !isImageValid ? 'border-red-500' : ''
              }`}
              onChange={(e) => handleInputChange('image', e.target.value)}
              onBlur={handleImageBlur}
              value={formData.image}
              required
            /> */}
            <input
              type="file"
              id="carPhoto"
              name="image"
              className={`border border-gray-300 rounded w-full ${!isImageValid ? 'border-red-500' : ''}`}
              onChange={(e) => handleImageChange('image', e.target.files[0])}
              onBlur={handleImageBlur}
              // value={formData.image}
              required
            />
            {!isImageValid && (
              <p className="text-red-500 text-sm">Please enter a valid image.</p>
            )}
          </div>

          <div className="mb-4">
            <label htmlFor="pricePerDay" className="block font-medium">
              Price Per Day:
            </label>
            <input
              type="number"
              id="pricePerDay"
              className="border border-gray-300 rounded px-3 py-2 w-full"
              onChange={(e) => handleInputChange('price', parseInt(e.target.value, 10))}
              value={formData.price}
              required
              min="0"
            />
          </div>

          <div className="mb-4">
            <label htmlFor="city" className="block font-medium">
              City:
            </label>
            <input
              type="text"
              id="city"
              className="border border-gray-300 rounded px-3 py-2 w-full"
              onChange={(e) => handleInputChange('city', e.target.value)}
              value={formData.city}
              required
            />
          </div>

          {error && <p className="text-red-500 text-sm">{error}</p>}

          <button
            type="submit"
            className="bg-lime-600 hover:bg-primary text-white rounded px-4 py-2"
          >
            <FontAwesomeIcon icon={faCar} className="pr-4" />
            Add Car
          </button>
        </form>
      </div>
    </section>
  );
};

export default AddCarPage;
