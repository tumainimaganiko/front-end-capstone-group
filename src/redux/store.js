import { configureStore } from '@reduxjs/toolkit';
import reservationReducer from './reservations/reservationSlice';
import carsReducer from './cars/carsSlice';
import modelsReducer from './cars/modelsSlice';

const store = configureStore({
  reducer: {
    reservations: reservationReducer,
    cars: carsReducer,
    models: modelsReducer,
  },
});

export default store;
