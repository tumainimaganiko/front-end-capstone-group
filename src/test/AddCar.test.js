import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../redux/store';
import AddCar from '../pages/AddCar';

it('Render a AddCar', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <MemoryRouter>
          <AddCar />
        </MemoryRouter>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
