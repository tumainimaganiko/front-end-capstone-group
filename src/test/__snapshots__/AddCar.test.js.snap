// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a AddCar 1`] = `
<section
  className="p-8"
>
  <div
    className="max-w-md mx-auto"
  >
    <h1
      className="text-2xl font-bold mb-4"
    >
      Add New Car
    </h1>
    <form
      id="Add-car-form"
      onSubmit={[Function]}
    >
      <div
        className="mb-4"
      >
        <label
          className="block font-medium"
          htmlFor="name"
        >
          Car Name:
        </label>
        <input
          className="border border-gray-300 rounded w-full "
          id="name"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="mb-4"
      >
        <label
          className="block font-medium"
          htmlFor="model_id"
        >
          Select Model:
        </label>
        <select
          className="border border-gray-300 rounded w-full"
          id="model_id"
          onChange={[Function]}
          required={true}
          value=""
        >
          <option
            value=""
          >
            -- Select a Model --
          </option>
        </select>
      </div>
      <div
        className="mb-4"
      >
        <label
          className="block font-medium"
          htmlFor="carDetails"
        >
          Car Plate:
        </label>
        <input
          className="border border-gray-300 rounded w-full "
          id="plate_number"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="mb-4"
      >
        <label
          className="block font-medium"
          htmlFor="carPhoto"
        >
          Car Photo:
        </label>
        <input
          className="border border-gray-300 rounded w-full "
          id="carPhoto"
          name="image"
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          type="file"
        />
      </div>
      <div
        className="mb-4"
      >
        <label
          className="block font-medium"
          htmlFor="pricePerDay"
        >
          Price Per Day:
        </label>
        <input
          className="border border-gray-300 rounded px-3 py-2 w-full"
          id="pricePerDay"
          min="0"
          onChange={[Function]}
          required={true}
          type="number"
          value=""
        />
      </div>
      <div
        className="mb-4"
      >
        <label
          className="block font-medium"
          htmlFor="city"
        >
          City:
        </label>
        <input
          className="border border-gray-300 rounded px-3 py-2 w-full"
          id="city"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
      <button
        className="bg-lime-600 hover:bg-primary text-white rounded px-4 py-2"
        type="submit"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-car pr-4"
          data-icon="car"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M135.2 117.4L109.1 192H402.9l-26.1-74.6C372.3 104.6 360.2 96 346.6 96H165.4c-13.6 0-25.7 8.6-30.2 21.4zM39.6 196.8L74.8 96.3C88.3 57.8 124.6 32 165.4 32H346.6c40.8 0 77.1 25.8 90.6 64.3l35.2 100.5c23.2 9.6 39.6 32.5 39.6 59.2V400v48c0 17.7-14.3 32-32 32H448c-17.7 0-32-14.3-32-32V400H96v48c0 17.7-14.3 32-32 32H32c-17.7 0-32-14.3-32-32V400 256c0-26.7 16.4-49.6 39.6-59.2zM128 288a32 32 0 1 0 -64 0 32 32 0 1 0 64 0zm288 32a32 32 0 1 0 0-64 32 32 0 1 0 0 64z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Add Car
      </button>
    </form>
  </div>
</section>
`;
